### PR TITLE
[#358] Port .Net Skill bots to Cloud Adapter - WaterfallHostBot

### PR DESCRIPTION
**Pull request by GiulianoDolceTalianoSantoro**
Addresses # 358

## Description
This PR updates the .NET WaterfallHostBot to use CloudAdapter.

### Detailed Changes
- Removed custom **_AllowedCallersClaimsValidator_** class in favor of the one from the SDK.
- Updated **_BotController_** constructors to receive _IBotFrameworkHttpAdapter_ instead of _BotFrameworkHttpAdapter_ as parameters.
- Updated **_AdapterWithErrorHandler_** class to extend `CloudAdapter` instead of `BotFrameworkHttpAdapter`.
- Updated **_LoggerMiddleware_** class "*_logger*" property type from `BotFrameworkHttpAdapter` to `CloudAdapter`.
- Updated **_Startup_** class to add a `BotFrameworkAuthentication` to the services. This is required by the Cloud Adapter constructor.
- Update **_TokenExchangeSkillHandler_** class to use _UserTokenClient_ instead of the _IExtendedUserTokenProvider_ for exchanging the token.
- Add **ComponentSettings** class for mapping _TokenExchange_ related properties.

## Testing
These images show the pipelines passing after the changes.

image

image

### CODE REVIEW

**Review by santgr11 — APPROVED** [review]

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/ComponentSettings.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
Comment (ceciliaavila) [inline Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/ComponentSettings.cs]: Please add the copyright header.

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/ComponentSettings.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
```
Comment (ceciliaavila) [inline Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/ComponentSettings.cs]: These lines are not needed.

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs [hunk 3]
```diff
@@ -29,29 +31,56 @@ public Startup(IConfiguration configuration)
...
+            services.AddSingleton<BotAdapter>(sp => sp.GetService<CloudAdapter>());
 
             // Register the skills conversation ID factory, the client and the request handler.
             services.AddSingleton<SkillConversationIdFactoryBase, SkillConversationIdFactory>();
-            services.AddHttpClient<SkillHttpClient>();
-            
-            //services.AddSingleton<ChannelServiceHandler, SkillHandler>();
+            services.AddHttpClient<SkillHttpClient>();
+
+            //services.AddSingleton<ChannelServiceHandler, SkillHandler>();
```
Comment (ceciliaavila) [inline Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs]: Can we remove this commented line?

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs [hunk 4]
```diff
@@ -29,29 +31,56 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddHttpClient();
```
Comment (ceciliaavila) [inline Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs]: Is this needed if we have `services.AddHttpClient<SkillHttpClient>()` below?

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/TokenExchangeSkillHandler.cs [hunk 5]
```diff
@@ -29,39 +29,37 @@ public class TokenExchangeSkillHandler : SkillHandler
```
Comment (ceciliaavila) [inline Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/TokenExchangeSkillHandler.cs:27]: In this class, there are some usings that are no longer needed. Also, the property WaterfallSkillBot can be deleted.

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/TokenExchangeSkillHandler.cs [hunk 6]
```diff
@@ -29,39 +29,37 @@ public class TokenExchangeSkillHandler : SkillHandler
...
+            _botAuth = botAuth;
+            _authConfig = authConfig;
             _conversationIdFactory = conversationIdFactory;
-            _logger = logger ?? NullLogger<TokenExchangeSkillHandler>.Instance;
+            _skillsConfig = skillsConfig ?? new SkillsConfiguration(configuration);
             _botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;
+            _logger = logger;
+
+            var settings = configuration.GetSection("Bot.Builder.Community.Components.TokenExchangeSkillHandler")?.Get<ComponentSettings>() ?? new ComponentSettings();
+            _connectionName = settings.TokenExchangeConnectionName ?? configuration.GetSection("SsoConnectionName")?.Value;
```
Comment (ceciliaavila) [inline Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/TokenExchangeSkillHandler.cs]: Should we update AppSettings.json to include this new section?

**Review by ceciliaavila — CHANGES_REQUESTED** [review]

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs (hunk 3 above)
Comment (GiulianoDolceTalianoSantoro) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs]: It was added by Josh, but I think we can

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/TokenExchangeSkillHandler.cs (hunk 6 above)
Comment (GiulianoDolceTalianoSantoro) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/TokenExchangeSkillHandler.cs]: I'll remove it and stay with how we used to get connection name before.

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs (hunk 3 above)
Comment (ceciliaavila) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs]: There's still a commented line here (81)

**Comment by sw-joelmut** [comment]
Feedback applied
https://dev.azure.com/joelmut1333/functional-tests/_build/results?buildId=2277&view=results

File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs (hunk 3 above)
Comment (sw-joelmut) [reply Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Startup.cs]: @ceciliaavila feedback applied, thanks

**Review by sw-joelmut — APPROVED** [review]

**Review by MartinLuccanera — APPROVED** [review]

**Review by ceciliaavila — APPROVED** [review]

**Comment by ceciliaavila** [comment]
Promoted to MS in PR# 479

**Pull request closed without merging**